File: backend/services/groqService.js
const Groq = require('groq-sdk');



/**
 * Get or create Groq client (singleton)
 */
const getGroqClient = () => {
  if (!process.env.GROQ_API_KEY) {
    throw new Error('GROQ_API_KEY is not set in environment variables');
  }
  return new Groq({ apiKey: process.env.GROQ_API_KEY });
};

/**
 * Call Groq LLM with a prompt
 * @param {string} systemPrompt - System instructions
 * @param {string} userPrompt - User message / prompt with context
 * @param {object} options - Optional overrides
 * @returns {Promise<string>} Raw LLM text response
 */
const callGroq = async (systemPrompt, userPrompt, options = {}) => {
  const client = getGroqClient();

  const {
    model = 'llama-3.3-70b-versatile',
    temperature = 0.3,
    maxTokens = 4096,
  } = options;

  try {
    const response = await client.chat.completions.create({
      model,
      temperature,
      max_tokens: maxTokens,
      messages: [
        {
          role: 'system',
          content: systemPrompt,
        },
        {
          role: 'user',
          content: userPrompt,
        },
      ],
      response_format: { type: 'json_object' },
    });

    const content = response.choices[0]?.message?.content || '';
    return content;
  } catch (error) {
    console.error(`❌ Groq API Error: ${error.message}`);
    throw new Error(`Groq API call failed: ${error.message}`);
  }
};

/**
 * Call Groq for array responses (JSON array, not object)
 * Falls back to json_object if array not supported
 */
const callGroqForArray = async (systemPrompt, userPrompt, options = {}) => {
  const client = getGroqClient();

  const {
    model = 'llama-3.3-70b-versatile',
    temperature = 0.3,
    maxTokens = 4096,
  } = options;

  // Wrap in object for json_object response format
  const wrappedSystemPrompt = systemPrompt + '\n\nIMPORTANT: Wrap your JSON array in an object: {"data": [...]}';

  try {
    const response = await client.chat.completions.create({
      model,
      temperature,
      max_tokens: maxTokens,
      messages: [
        { role: 'system', content: wrappedSystemPrompt },
        { role: 'user', content: userPrompt },
      ],
      response_format: { type: 'json_object' },
    });

    const content = response.choices[0]?.message?.content || '{"data":[]}';
    const parsed = JSON.parse(content);
    return parsed.data || parsed;
  } catch (error) {
    console.error(`❌ Groq Array API Error: ${error.message}`);
    throw new Error(`Groq API call failed: ${error.message}`);
  }
};

module.exports = { callGroq, callGroqForArray };

File: ai-service/services/groq_service.py
"""
Groq Service — LLM Inference via Groq API
==========================================
Provides a unified interface for calling the Groq Llama model.
"""

import os
import json
from groq import Groq
from dotenv import load_dotenv

load_dotenv()

_client: Groq | None = None


def get_groq_client() -> Groq:
    """Get or create the Groq client (singleton)."""
    global _client
    if _client is None:
        api_key = os.getenv("GROQ_API_KEY")
        if not api_key:
            raise ValueError("GROQ_API_KEY is not set in environment variables")
        _client = Groq(api_key=api_key)
    return _client


async def call_groq(
    system_prompt: str,
    user_prompt: str,
    model: str | None = None,
    temperature: float = 0.3,
    max_tokens: int = 4096,
) -> str:
    """
    Call the Groq LLM and return the raw text response.

    Args:
        system_prompt: System instructions (role + behavior)
        user_prompt:   User message with context
        model:         Groq model ID (defaults to env var)
        temperature:   Sampling temperature
        max_tokens:    Max tokens to generate

    Returns:
        Raw LLM text response string
    """
    client = get_groq_client()
    model = model or os.getenv("GROQ_MODEL", "llama-3.3-70b-versatile")

    response = client.chat.completions.create(
        model=model,
        temperature=temperature,
        max_tokens=max_tokens,
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
        response_format={"type": "json_object"},
    )

    return response.choices[0].message.content or ""


async def call_groq_for_list(
    system_prompt: str,
    user_prompt: str,
    model: str | None = None,
    temperature: float = 0.3,
    max_tokens: int = 4096,
) -> list:
    """
    Call Groq for array-style responses.
    Wraps the array in an object since Groq requires json_object format.

    Returns:
        Python list from the LLM response
    """
    wrapped_system = system_prompt + '\n\nIMPORTANT: Wrap your JSON array in an object: {"data": [...]}'
    raw = await call_groq(wrapped_system, user_prompt, model, temperature, max_tokens)

    try:
        parsed = json.loads(raw)
        return parsed.get("data", parsed) if isinstance(parsed, dict) else parsed
    except json.JSONDecodeError:
        return []
